tests/suites/test_suite_psa_crypto.function

```diff
@@ -264,6 +264,334 @@ typedef enum {
+                }
+            }
+
+            PSA_ASSERT( psa_aead_update_ad( &operation,
+                                            additional_data->x + part_offset,
+                                            part_length ) );
+
+            part_offset += part_length;
+        }
+    }
+    else
+    {
+        /* Pass additional data in one go. */
+        PSA_ASSERT( psa_aead_update_ad( &operation, additional_data->x,
+                                        additional_data->len ) );
+    }
+
+    if( data_part_len != -1 )
+    {
+        /* Pass data in parts */
+        part_data_size = PSA_AEAD_UPDATE_OUTPUT_SIZE( key_type, alg,
+                                                      ( size_t ) data_part_len );
+
+        ASSERT_ALLOC( part_data, part_data_size );
+
+        part_offset = 0;
+
+        while( part_offset < data_true_size )
+        {
+            if( do_zero_parts && part_count++ & 0x01 )
+            {
+                part_length = 0;
+            }
+            else
+            {
+                if( ( data_true_size - part_offset ) < ( uint32_t ) data_part_len )
+                {
+                    part_length = ( data_true_size - part_offset );
+                }
+                else
+                {
+                    part_length = data_part_len;
+                }
+            }
+
+            PSA_ASSERT( psa_aead_update( &operation,
+                                         ( input_data->x + part_offset ),
+                                         part_length, part_data,
+                                         part_data_size,
+                                         &output_part_length ) );
+
+            if( output_data && output_part_length )
+            {
+                memcpy( ( output_data + part_offset ), part_data,
+                        output_part_length );
+            }
+
+            part_offset += part_length;
+            output_length += output_part_length;
+        }
+    }
+    else
+    {
+        /* Pass all data in one go. */
+        PSA_ASSERT( psa_aead_update( &operation, input_data->x,
+                                     data_true_size, output_data,
+                                     output_size, &output_length ) );
+    }
+
+    if( is_encrypt )
+        PSA_ASSERT( psa_aead_finish( &operation, final_data,
+                                     final_output_size,
+                                     &output_part_length,
+                                     tag_buffer, tag_length,
+                                     &tag_size ) );
+    else
+    {
+        status = psa_aead_verify( &operation, final_data,
+                                     final_output_size,
+                                     &output_part_length,
+                                     ( input_data->x + data_true_size ),
+                                     tag_length );
+
+        if( status != PSA_SUCCESS )
+        {
+            if( !expect_valid_signature )
```

Comment:
That commit was primarily because of a comment from yourself about how the expected failures were only invalid signature, and thus expected_status was overkill. There is also the problem of which call to the api you are expecting to return an error, as I was having to check all returns to see if they were invalid and then checking them against the expected result, which wasn't particularly nice.

Reply:
I think that was in an earlier version of this function which allowed any function to error out. With the current version, where only verify might fail, I think `expect_valid_signature` is an unnecessary complication.

And now that I look again, it actually causes a bug. Whether to use `expected_status` or `expect_valid_signature` is not a blocker, but this bug is a blocker. The bug is that if `psa_aead_verify` returns `PSA_SUCCESS` but the test data expected a verification failure, the test will pass, hiding a bug in the library. A simple `TEST_EQUAL( actual_status, expected_status )` wouldn't have this bug.